Documentation/Coherency.md

```diff
@@ -0,0 +1,74 @@
+
+# Solutions Considered #
+------------------------
+We have considered a few solutions:
+
+## Continue with traditional source-build coherence ##
+This would mean that we would continue to force coherence on every build, using our traditional method of passing around PackageVersions.props as the build progresses.
+Pros:
+- Simple.
+- Already have this mechanism.
+
+Cons:
+- We continue to not ship bits that match official builds, except on releases.
+- We have to deal with the problem of not getting all packages for servicing releases:
+	-  Issues https://github.com/dotnet/source-build/issues/210 and https://github.com/dotnet/source-build/issues/923 are our earlier discussions.
+	-  If we turn on BuildAllConfigurations and BuildAllPackages, we will ship bits that do not exist except on RHEL machines.  If someone shares a project but not the packages directory, restoring these packages that do not officially exist will fail.
+	-  If we don't turn on BuildAllConfigurations and BuildAllPackages, we will continue to have a large number of prebuilt packages.
+
+## Build multiple versions of each repo ##
+This would mean we build CLI with its expected version of CoreFX, and core-setup with its own expected version of CoreFX.
+Pros:
+- We match the official bits very closely.
+- We don't have to deal with servicing considerations.
+
+Cons:
+- This explodes rather quickly and would have extremely long build times.
+- We have to be careful about cross-contaminating package feeds during the builds.
+
+## Store and download packages in later builds ##
+During each build, we would save source-built packages off in an Azure account or NuGet feed.  During later builds, we would restore these packages and use them in builds later in the dependency tree.
+Pros:
+- Improved build times.
+	- Fringe benefit: incremental builds would probably finally work.
+- Still lets us eliminate prebuilt packages (all packages stored are source-built).
+- We match the official bits very closely.
+
+Cons:
+- Have to be very careful to not contaminate the storage with prebuilt packages.
+- Need internet access during the build.
+- Probably does not satisfy requirements for no prebuilt binaries.
+
+# Proposed Solution #
+---------------------
+Hybrid approach: we will enable storing and retrieving packages, but allow building multiple versions of each repo completely from source.  Daily builds and dev builds will mostly use the stored packages approach while release builds will use the completely-built-from-source approach.  Doing multiple versions of repos only for release builds should cut down on the numbers of builds that are done many times since these builds should be mostly coherent except for servicing considerations.
```

Comment:
> Daily builds and dev builds will mostly use the stored packages approach while release builds will use the completely-built-from-source approach.

From the perspective of a distro maintainer, it doesn't seem maintainable for servicing release build times to be O(n) based on the number of releases that have happened. I think we need to find the way to store these packages for future builds to depend on for this strategy to work.

Reply:
Adding a section for publishing actual packages instead of as build intermediates.